src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala

```diff
@@ -391,63 +406,78 @@ class Checker(globalTable: SymbolTable, verbose: Boolean = false) {
     private def inferAndCheckExpr(decl: InvokableDeclaration,
                                   context: Context,
                                   e: Expression,
-                                  ownershipConsumptionMode: OwnershipConsumptionMode): (ObsidianType, Context) = {
+                                  ownershipConsumptionMode: OwnershipConsumptionMode): (ObsidianType, Context, Expression) = {
 
         /* returns [t] if [e : t], otherwise returns BottomType */
-        def assertTypeEquality(e: Expression, t: ObsidianType, c: Context): (ObsidianType, Context) = {
-            val (tPrime, contextPrime) = inferAndCheckExpr(decl, c, e, NoOwnershipConsumption())
-            (checkIsSubtype(e, tPrime, t), contextPrime)
+        def assertTypeEquality(e: Expression, t: ObsidianType, c: Context): (ObsidianType, Context, Expression) = {
+            val (tPrime, contextPrime, ePrime) = inferAndCheckExpr(decl, c, e, NoOwnershipConsumption())
+            (checkIsSubtype(e, tPrime, t), contextPrime, ePrime)
         }
 
-        def assertOperationType(e1: Expression, e2: Expression, t: ObsidianType): (ObsidianType, Context) = {
-            val (_, c1) = assertTypeEquality(e1, t, context)
-            val (_, c2) = assertTypeEquality(e2, t, c1)
-            (t, c2)
+        def assertOperationType(e1: Expression, e2: Expression, t: ObsidianType): (ObsidianType, Context, Expression, Expression) = {
+            val (_, c1, e1Prime) = assertTypeEquality(e1, t, context)
+            val (_, c2, e2Prime) = assertTypeEquality(e2, t, c1)
+            (t, c2, e1Prime, e2Prime)
         }
 
-        def assertComparisonType(e1: Expression, e2: Expression): (ObsidianType, Context) = {
-            val (_, c1) = assertTypeEquality(e1, IntType(), context)
-            val (_, c2) = assertTypeEquality(e2, IntType(), c1)
-            (BoolType(), c2)
+        def assertComparisonType(e1: Expression, e2: Expression): (ObsidianType, Context, Expression, Expression) = {
+            val (_, c1, e1Prime) = assertTypeEquality(e1, IntType(), context)
+            val (_, c2, e2Prime) = assertTypeEquality(e2, IntType(), c1)
+            (BoolType(), c2, e1Prime, e2Prime)
         }
 
         def handleInvocation(
```

Comment:
That's a lot of outputs. Please add a comment explaining what they are (or refactor so there aren't so many).